### PR TITLE
Obtain the maximum value of a two-dimensional discontinuous tensor without performing time-consuming continuous operations

### PR DESCRIPTION
**Pull request by haowen-han**
测试脚本:
```
import torch
import flag_gems
torch.manual_seed(1)
for i in range(10):
    A = torch.randn((4096, 2560), dtype=torch.bfloat16, device="cuda")[::2,::2]

    with flag_gems.use_gems():
        triton_out = torch.max(A)
import time

A = torch.randn((4096, 2560), dtype=torch.bfloat16, device="cuda")[::2,::2]
with flag_gems.use_gems():
    start_time = time.time()
    for _ in range(10000):
        triton_out = torch.max(A)
    gems_time = time.time() - start_time

print(f"shape:{A.shape},gems_time:{gems_time}")

A = torch.randn((4096, 4096), dtype=torch.bfloat16, device="cuda")[::2,::2]
with flag_gems.use_gems():
    start_time = time.time()
    for _ in range(10000):
        triton_out = torch.max(A)
    gems_time = time.time() - start_time

print(f"shape:{A.shape},gems_time:{gems_time}")

A = torch.randn((3053, 2560), dtype=torch.bfloat16, device="cuda")[::2,::2]
with flag_gems.use_gems():
    # 使用GEMS的时间
    start_time = time.time()
    for _ in range(10000):
        triton_out = torch.max(A)
    gems_time = time.time() - start_time

print(f"shape:{A.shape},gems_time:{gems_time}")


A = torch.randn((6013, 5024), dtype=torch.bfloat16, device="cuda")[::2,::2]
with flag_gems.use_gems():
    start_time = time.time()
    for _ in range(10000):
        triton_out = torch.max(A)
    gems_time = time.time() - start_time

print(f"shape:{A.shape},gems_time:{gems_time}")

A = torch.randn((331, 532), dtype=torch.bfloat16, device="cuda")[::2,::2]
with flag_gems.use_gems():
    start_time = time.time()
    for _ in range(10000):
        triton_out = torch.max(A)
    gems_time = time.time() - start_time

print(f"shape:{A.shape},gems_time:{gems_time}")

```
原来的时间：
企业微信截图_17316689691077
本pr的时间：
企业微信截图_17316688991443
测试环境:A100 80GB,考虑到for循环次数较多，因此没有添加LAUNCH_BLOCKING环境变量

### CODE REVIEW

**Comment by haowen-han** [comment]
> please mention the issue this pr solves:)

no issue, just want to accelerate..

**Comment by iclementine** [comment]
Good idea. We planned to implement a codegen-based reduction kernel that can handle arbitary rank and stride. It involves kernel that can handle strides, and some analysis on shape and stride to combine some dimensions. It may have similar logic with TensorIterator on torch.

**Comment by haowen-han** [comment]
> Good idea. We planned to implement a codegen-based reduction kernel that can handle arbitary rank and stride. It involves kernel that can handle strides, and some analysis on shape and stride to combine some dimensions. It may have similar logic with TensorIterator on torch.

so will this pr be accepted? I think maybe we can delete it after iterator is done, but now we can accept it.

**Comment by iclementine** [comment]
We would continue this PR as usual.

File: src/flag_gems/ops/max.py [hunk 1]
```diff
@@ -91,8 +91,91 @@ def max_kernel(
...
+        block_mid = triton.next_power_of_2(inp.shape[0])
+        max_kernel_2[(1, 1, 1)](mid, out, inp.shape[0], block_mid)
+        return out
+
+
+def max(inp: torch.Tensor):
     logging.debug("GEMS MAX")
+    if inp.ndim == 2 and inp.is_contiguous() is False:
+        out = max_2d_uncontiguous(inp)
+        return out
```
Comment (iclementine) [inline src/flag_gems/ops/max.py:178]: I see. This specialized implementation is an optimization for reduction on all axes.(reduction to scalar)

File: src/flag_gems/ops/max.py [hunk 2]
```diff
@@ -91,8 +91,91 @@ def max_kernel(
...
+    inp_shape1,
+    mid_ptr,
+    BLOCK_SIZE: tl.constexpr,
+):
+    block_idx = tl.program_id(0)
+    block_nums = tl.num_programs(0)
+
+    number_of_row_to_process = tl.cdiv(inp_shape0, block_nums)
+    start_row = block_idx * number_of_row_to_process
+    for row_index in range(number_of_row_to_process):
```
Comment (iclementine) [inline src/flag_gems/ops/max.py:134]: Did you try the alternative partition scheme like this


图片

Assume there are 5 CTAs and 20 rows here, and each CTA takes 4 rows, but in a circular order. In this case, consecutively scheduled CTAs reads adjacent lines.

File: tests/test_general_reduction_ops.py [hunk 3]
```diff
@@ -124,10 +124,21 @@ def test_accuracy_max_without_dim(shape, dtype):
...
+    inp = torch.randn(shape, dtype=dtype, device="cuda")[..., ::2, ::2]
+    ref_inp = to_reference(inp)
+
+    ref_out = torch.max(ref_inp)
+    with flag_gems.use_gems():
+        res_out = torch.max(inp)
+
+    gems_assert_equal(res_out, ref_out)
+
+    inp = torch.randn(shape, dtype=dtype, device="cuda")[..., ::3, ::5]
```
Comment (iclementine) [inline tests/test_general_reduction_ops.py:141]: Did you test cases where the input is unique and non-overlapping but not C-contiguous? Did this PR improve performance at these cases?

**Pull request closed without merging**